### ISSUE TITLE
Make DatabaseReplicated multithreaded

### ISSUE THREAD

**Issue by alesapin** [ClickHouse/ClickHouse#56690]
`DatabaseReplicated` replicate DLL queries (`CREATE`, `ALTER`, `DROP`, `RENAME`, `EXCHANGE`, `SYNC REPLICA`) across multiple replicas. Initially user query creates DDLEntry in DatabaseReplicated replication queue and then all the replicas pick entries one by one in sequential order, execute them and confirm execution in ZooKeeper. 

In case of large amount of replicas and DDL queries (a lot of tables, regular `ALTERs` and so on) the execution of queue can take a long time. The proposal is to execute independent queries in parallel. 

In general case it's really hard to precisely identify independent queries, some examples:
1. `CREATE TABLE test AS SELECT * FROM some_other_table`
2. `ALTER TABLE test UPDATE column x = (SELECT str FROM string_table LIMIT 1) WHERE y IN (SELECT id FROM id_table)`
So instead of precise information about all qualified identifiers we need to collect the following:
1. *ProduceSet:* Set of identifiers we produce/change. In this set we store identifiers which can be obtained from any DDL query with `ASTQueryWithTableAndOutput::getTable()` + some special code for `ASTRenameQuery`. This set is small and precise. 
2. *UseSet:* Set of identifiers we touch/use. Here we should be as pessimistic as possible: just collect all used identifiers with `IdentifiersCollector` + special code for `CREATE` queries (`getDependenciesFromCreateQuery`).

Using these two sets we can track dependencies between queries in the following way. Query can be scheduled for concurrent execution when:
* All queries where `ProduceSet` intersects with our `UseSet` are not executed concurrently.
* All queries where `UseSet` intersects with our `ProduceSet` are not executed concurrently.
All queries are scheduled for execution in order of their numbers in queue. We don't intentionally reorder them. So the execution loop should look like:
```
for (entry : ordered_entries)
{
    if (entry.use_set.intersection(currently_executing_produce_set).empty() 
        && entry.produce_set.intersection(currently_executing_use_set).empty()) {
        scheduleEntry(entry); /// fill both sets at start and cleanup both sets at the end of execution
    } else {
        /// postpone (more complex) or wait (more simple). 
    }
}
```

`DDLWorker` already supports multi-threaded thread-pool execution, however without any dependencies checks. 

*Caveats:*
`DatabaseReplicatedWorker` (extension of `DDLWorker`) has `log_pointer` for each replica. This `log_pointer` is mostly used to log shrinking. With concurrent entries execution we should update in a little bit more complex way.

**Comment by alesapin** [comment]
cc @tavplubix

**Comment by alexey-milovidov** [comment]
We can simplify it and collect a set of identifiers from a query.
If queries have at least a common identifier, they should be processed in order, otherwise in parallel.
The only caveat is that we have to also collect table names from string literals and do constant folding.

**Comment by al13n321** [comment]
Does this work correctly with views? E.g. the following sequence:
```
1. create table a (x Int8) engine Memory
2. insert into a values (1)
3. create view v as select * from a
4. alter table a update x=2 where x=1
5. create table t engine Memory as select * from v
```

If queries 4 and 5 are reordered, the result will change, but this dependency can't be inferred from just the ASTs.

(If this reordering seems ok, replace query 4 with dropping+creating `a`. Then query 5 will fail if it runs at the wrong time, while `a` doesn't exist. Seems clearly bad.)

---

Can we just say that if a query depends on any views then we don't run it in parallel with anything else? In general, no: at the time of analysis, the view may not exist, or the name may refer to a regular table that later gets replaced with a view.

We could recurse into views when collecting dependencies, but that also seems tricky:
 * The view may not have been created yet. In example above, if we try to collect dependencies for query 5 before we executed query 3, it won't work. So the collection of dependencies would also have to go through some dependency resolution mechanism. Can be handled by the same loop: if we failed to form the `use_set` because some tables don't exist, postpone, then later retry the analysis. (When exactly to retry? I guess in case of failure we should still get a partial `use_set`, and we should retry when all of those dependencies are met. Or just retry after each executed query - that's at most 2 analyses per query, amortized, if we imagine that the query that finished execution "pays" for the re-analysis of the query at the front of the queue.)
 * The view may be dropped+created between the analysis and the execution. Or it could be a regular table at the time of analysis, then turn into a view later. I guess this is solved by just re-analyzing the query before executing it.
 * Maybe `getDependenciesFromCreateQuery` would have to be more careful and less best-effort, to avoid stalling the DDL loop with false dependencies. Or maybe it's not a big deal: the query will just unnecessarily wait for all previous queries to complete, then we'll run it because it's the first in queue, regardless of its dependencies. Even if that happens for every query, it's no worse than what we have today.

**Comment by alesapin** [comment]
@al13n321 The good news is that `CREATE ... AS SELECT * FROM xxx` allowed for database replicated only in form of `CREATE ... AS SELECT * FROM xxx EMPTY`.

**Comment by tavplubix** [comment]
And the bad news are that we allow subqueries in alter update/delete

**Comment by tavplubix** [comment]
And the good news is that alter update/delete doesn't go through the DDL queue in Replicated database if we have only one shard and the table is replicated :)

**Comment by alesapin** [comment]
Parallel execution require to change the logic of `log_ptr` and `max_log_ptr`. All entries in DDLWoker queue are self-explainable. They contain records about who executed them already, what is currently executing and so on. Why we need log pointers at all? Three reasons:
1. `log_ptr` -- To understand that our replica is stale and we removed some entries from log which were not executed by this replica.
2. `max_log_ptr` -- To collect consistent metadata snapshot during replica recovery.
3. `max_log_ptr` -- To understand which records we can remove.

If we introduce concurrent log execution we are in trouble, because both `max_log_ptr` and `log_ptr` can have almost random value now. Semantically `log_ptr` should become "the first undone entry on this replica" and `max_log_ptr` the `max(log_ptr)` across all replicas. 
1) Will work in the same way, we are stale if we didn't execute some entry which could be already removed.
2) It will change. Get consistent snapshot and max_log_ptr. However we will have to avoid execution of all entries which we executed by at least one replica. So we will have to remember that we shouldn't execute them in some way.
3) Will work in the same way, we can remove all old records if there are at least one replica who executed all of them.

Technical implementation:
Each replica has own node `/finished_tasks`. Which contain finished tasks in children list. After task was executed on replica it do the following:
```
if (log_ptr == our_task_number)
{
    sorted_finished_tasks, version = sort(getChildren(finished_tasks));
    assert(sorted_finished_tasks.empty() || our_task < sorted_finished_tasks[0]);
    sorted_finished_tasks.push_front(our_task);
    prev_finished_task = our_task;
    for (i = 0; i < sorted_finished_tasks.size() ++i)
    {
        if (sorted_finished_tasks[i] - prev_finished_task > 1)
            break;
       prev_finished_task = sorted_finished_tasks[i];
    }
    tx.remove(sorted_finished_tasks[1:i]);
    tx.set(log_ptr, sorted_finished_tasks[i] + 1);
    tx.cas(max_log_ptr, sorted_finished_tasks[i] + 1) // compare and set if greater
    tx.check(finished_tasks, version);
}
else 
{
    tx.create(/finished_tasks/our_task_number)
    tx.set(/finished_tasks, "") // increment version
}
```
For snapshot we should take not only `max_log_ptr`  and tables state, but also queue state. To verify that it's consistent each replica on task finish should update version of `log`|`queue` node. The code will look like:
```
while (true)
{
    log_version = get_version(/log)
    max_log_ptr = get(/max_log_ptr)
    state = foreach table in getChildren(state)
    queue = foreach queue_entry in getChildren(queue) 
    if (check_version(log, log_version))
        break;
}
recover(log_version, state, queue);
for (entry : queue)
{
    if (entry.finishedBySomeone())
        tx.create(entry.path / finished / host_id);
}
tx.create(log_ptr, max_log_ptr);
```

**Comment by alexey-milovidov** [comment]
No plans for this anymore.
